core/kernel/tee_ta_manager.c

```diff
@@ -277,6 +277,33 @@ static TEE_Result check_client(struct tee_ta_session *s, const TEE_Identity *id)
 	return TEE_SUCCESS;
 }
 
+// TODO: move this to core/arch/arm/user_ta.c|static_ta.c ? */
+#include <mm/mobj.h>
+static TEE_Result check_params(struct tee_ta_session *s,
+				struct tee_ta_param *param)
+{
+	int n;
+
+	/* check that non SDP TAs do not get SDP shm memrefs */
+	if (s->ctx->flags & TA_FLAG_SECURE_DATA_PATH)
+		return TEE_SUCCESS;
+
+	for (n = 0; n < TEE_NUM_PARAMS; n++) {
+		uint32_t param_type = TEE_PARAM_TYPE_GET(param->types, n);
+		struct param_mem *mem = &param->u[n].mem;
+
+		if (param_type != TEE_PARAM_TYPE_MEMREF_INPUT &&
+		    param_type != TEE_PARAM_TYPE_MEMREF_OUTPUT &&
+		    param_type != TEE_PARAM_TYPE_MEMREF_INOUT)
+			continue;
+		if (!mem->size)
+			continue;
+		if (mobj_matches(mem->mobj, CORE_MEM_SDP_SHM))
+			return TEE_ERROR_BAD_PARAMETERS;
+	}
+	return TEE_SUCCESS;
+}
```

Comment:
Yes, that's a pity, not much we can do about that. Optimizations etc should preferably be done in a later pull request. No further checks should be added, one nice thing with the MOBJs is that the fact that you have a MOBJ guarantees that it's OK (except in the SDP case...).